File: CS/CustomSecurityOperation.Module/CustomSecurityObjects.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using DevExpress.ExpressApp.Security;
using DevExpress.ExpressApp.DC;
using DevExpress.ExpressApp;
using DevExpress.Xpo;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl.PermissionPolicy;

namespace CustomSecurityOperation.Module {
    public class ExportPermission : IOperationPermission {
        public ExportPermission(Type objectType, SecurityPermissionState state) {
            ObjectType = objectType;
            State = state;
        }
        public Type ObjectType { get; private set; }

        public string Operation {
            get {
                return "Export";
            }
        }
        public SecurityPermissionState State { get; set; }
    }
    public class ExportPermissionRequest : IPermissionRequest {
        public ExportPermissionRequest(Type objectType) {
            ObjectType = objectType;
        }
        public Type ObjectType { get; private set; }
        public object GetHashObject() {
            return this.GetType().FullName;
        }
    }

    public class ExportPermissionRequestProcessor : PermissionRequestProcessorBase<ExportPermissionRequest> {
        private IPermissionDictionary permissionDictionary;
        public ExportPermissionRequestProcessor(IPermissionDictionary permissionDictionary) {
            this.permissionDictionary = permissionDictionary;
        }
        public override bool IsGranted(ExportPermissionRequest permissionRequest) {
            IEnumerable<ExportPermission> exportPermissions = permissionDictionary.GetPermissions<ExportPermission>().Where(p => p.ObjectType == permissionRequest.ObjectType);
            if (exportPermissions.Count() == 0) {
                return IsGrantedByPolicy(permissionDictionary);
            }
            else {
                return exportPermissions.Any(p => p.State == SecurityPermissionState.Allow);
            }
        }

        private bool IsGrantedByPolicy(IPermissionDictionary permissionDictionary) {
            if (GetPermissionPolicy(permissionDictionary) == SecurityPermissionPolicy.AllowAllByDefault) {
                return true;
            }
            return false;
        }

        private SecurityPermissionPolicy GetPermissionPolicy(IPermissionDictionary permissionDictionary) {
            SecurityPermissionPolicy result = SecurityPermissionPolicy.DenyAllByDefault;
            List<SecurityPermissionPolicy> permissionPolicies = permissionDictionary.GetPermissions<PermissionPolicy>().Select(p => p.SecurityPermissionPolicy).ToList();
            if (permissionPolicies != null && permissionPolicies.Count != 0) {
                if (permissionPolicies.Any(p => p == SecurityPermissionPolicy.AllowAllByDefault)) {
                    result = SecurityPermissionPolicy.AllowAllByDefault;
                }
                else {
                    if (permissionPolicies.Any(p => p == SecurityPermissionPolicy.ReadOnlyAllByDefault)) {
                        result = SecurityPermissionPolicy.ReadOnlyAllByDefault;
                    }
                }
            }
            return result;
        }
    }
    [XafDisplayName("Type Operation Permissions")]
    public class CustomTypePermissionObject : PermissionPolicyTypePermissionObject {
        public CustomTypePermissionObject(Session session)
            : base(session) {
        }
        [XafDisplayName("Export")]
        public SecurityPermissionState? ExportState {
            get {
                return GetPropertyValue<SecurityPermissionState?>("ExportState");
            }
            set {
                SetPropertyValue("ExportState", value);
            }
        }
    }
}

File: VB/CustomSecurityOperation.Module/CustomSecurityObjects.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Linq
Imports DevExpress.ExpressApp.Security
Imports DevExpress.ExpressApp.DC
Imports DevExpress.ExpressApp
Imports DevExpress.Xpo
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl.PermissionPolicy

Namespace CustomSecurityOperation.Module
	Public Class ExportPermission
		Implements IOperationPermission
		Public Sub New(ByVal objectType As Type, ByVal state As SecurityPermissionState)
			Me.ObjectType = objectType
			Me.State = state
		End Sub
		Private privateObjectType As Type
		Public Property ObjectType() As Type
			Get
				Return privateObjectType
			End Get
			Private Set(ByVal value As Type)
				privateObjectType = value
			End Set
		End Property
		Public ReadOnly Property Operation() As String Implements IOperationPermission.Operation
			Get
				Return "Export"
			End Get
		End Property
		Public Property State() As SecurityPermissionState
	End Class
	Public Class ExportPermissionRequest
		Implements IPermissionRequest
		Public Sub New(ByVal objectType As Type)
			Me.ObjectType = objectType
		End Sub
		Private privateObjectType As Type
		Public Property ObjectType() As Type
			Get
				Return privateObjectType
			End Get
			Private Set(ByVal value As Type)
				privateObjectType = value
			End Set
		End Property
		Public Function GetHashObject() As Object Implements IPermissionRequest.GetHashObject
			Return ObjectType.FullName
		End Function
	End Class

	Public Class ExportPermissionRequestProcessor
		Inherits PermissionRequestProcessorBase(Of ExportPermissionRequest)
		Private permissionDictionary As IPermissionDictionary
		Public Sub New(ByVal permissionDictionary As IPermissionDictionary)
			Me.permissionDictionary = permissionDictionary
		End Sub
		Public Overrides Function IsGranted(ByVal permissionRequest As ExportPermissionRequest) As Boolean
			Dim exportPermissions As IEnumerable(Of ExportPermission) = permissionDictionary.GetPermissions(
		Of ExportPermission)().Where(Function(p) p.ObjectType = permissionRequest.ObjectType)
			If exportPermissions.Count() = 0 Then
				Return IsGrantedByPolicy(permissionDictionary)
			Else
				Return exportPermissions.Any(Function(p) p.State = SecurityPermissionState.Allow)
			End If
		End Function
		Private Function IsGrantedByPolicy(ByVal permissionDictionary As IPermissionDictionary) As Boolean
			If GetPermissionPolicy(permissionDictionary) = SecurityPermissionPolicy.AllowAllByDefault Then
				Return True
			End If
			Return False
		End Function
		Private Function GetPermissionPolicy(ByVal permissionDictionary As IPermissionDictionary) As SecurityPermissionPolicy
			Dim result As SecurityPermissionPolicy = SecurityPermissionPolicy.DenyAllByDefault
			Dim permissionPolicies As List(Of SecurityPermissionPolicy) = permissionDictionary.GetPermissions(
		Of PermissionPolicy)().Select(Function(p) p.SecurityPermissionPolicy).ToList()
			If permissionPolicies IsNot Nothing AndAlso permissionPolicies.Count <> 0 Then
				If permissionPolicies.Any(Function(p) p = SecurityPermissionPolicy.AllowAllByDefault) Then
					result = SecurityPermissionPolicy.AllowAllByDefault
				Else
					If permissionPolicies.Any(Function(p) p = SecurityPermissionPolicy.ReadOnlyAllByDefault) Then
						result = SecurityPermissionPolicy.ReadOnlyAllByDefault
					End If
				End If
			End If
			Return result
		End Function
	End Class
	<XafDisplayName("Type Operation Permissions")>
	Public Class CustomTypePermissionObject
		Inherits PermissionPolicyTypePermissionObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		<XafDisplayName("Export")>
		Public Property ExportState() As SecurityPermissionState?
			Get
				Return GetPropertyValue(Of SecurityPermissionState?)(NameOf(ExportState))
			End Get
			Set(ByVal value? As SecurityPermissionState)
				SetPropertyValue(NameOf(ExportState), value)
			End Set
		End Property
	End Class
End Namespace
